File: functions/src/musicCompatibility.ts
export const MUSIC_SCOPES =
  "user-top-read user-read-recently-played playlist-read-private";
export const SYNC_MIN_INTERVAL_MS = 6 * 60 * 60 * 1000;
export const MUSIC_RANKING_WEIGHT = 0.15;
export const MUSIC_PROFILE_VERSION = 2;
/** MVP: last N unique recently-played Spotify track IDs (deduped by track id). */
export const RECENT_UNIQUE_TRACK_LIMIT = 10;

/** Default weights when playlist overlap data is unavailable. */
export const MUSIC_WEIGHTS_BASE = {
  tracks: 0.4,
  artists: 0.3,
  genres: 0.2,
  recent: 0.1,
  playlist: 0,
} as const;

/** Weights when both users have playlist track IDs from Spotify. */
export const MUSIC_WEIGHTS_WITH_PLAYLIST = {
  tracks: 0.35,
  artists: 0.25,
  genres: 0.2,
  playlist: 0.1,
  recent: 0.1,
} as const;

export type MusicTaste = {
  trackIds: string[];
  artistIds: string[];
  genres: string[];
  recentTrackIds: string[];
  recentArtistIds: string[];
  /** Minimized playlist track IDs only — never fabricated play counts. */
  playlistTrackIds?: string[];
};

export type MusicCompatibilityBreakdown = {
  tracks: number;
  artists: number;
  genres: number;
  recent: number;
  playlist: number;
};

export type MusicInsightCode =
  | "band_very_high"
  | "band_high"
  | "band_mid"
  | "band_low"
  | "shared_tracks"
  | "shared_artists"
  | "shared_playlist_tracks"
  | "shared_recent_tracks"
  | "top_shared_artist"
  | "top_shared_genres"
  | "data_unavailable";

export type MusicInsight = {
  code: MusicInsightCode;
  /** Optional display params (names / counts). Never fake listening stats. */
  params?: Record<string, string | number>;
};

export type MusicCompatibility = {
  score: number;
  sharedTracks: string[];
  sharedArtists: string[];
  sharedGenres: string[];
  sharedRecentTracks: string[];
  sharedPlaylistTracks: string[];
  breakdown: MusicCompatibilityBreakdown;
  insights: MusicInsight[];
};

export function isTasteEmpty(taste: MusicTaste | null | undefined): boolean {
  if (!taste) return true;
  return (
    taste.trackIds.length === 0 &&
    taste.artistIds.length === 0 &&
    taste.genres.length === 0 &&
    taste.recentTrackIds.length === 0 &&
    taste.recentArtistIds.length === 0 &&
    (taste.playlistTrackIds?.length ?? 0) === 0
  );
}

function norm(value: string): string {
  return value.trim().toLowerCase();
}

function intersect(a: string[], b: string[]): string[] {
  const other = new Set(b.map(norm).filter((item) => item.length > 0));
  const seen = new Set<string>();
  const out: string[] = [];
  for (const item of a) {
    const key = norm(item);
    if (!key || !other.has(key) || seen.has(key)) continue;
    seen.add(key);
    out.push(item.trim());
  }
  return out;
}

function overlap(a: string[], b: string[]): number {
  if (!a.length || !b.length) return 0;
  const shared = intersect(a, b).length;
  const denom = Math.min(a.length, b.length);
  return denom === 0 ? 0 : Math.min(1, shared / denom);
}

function recentOverlap(viewer: MusicTaste, candidate: MusicTaste): {
  ratio: number;
  sharedTracks: string[];
  sharedArtists: string[];
} {
  const sharedTracks = intersect(viewer.recentTrackIds, candidate.recentTrackIds);
  const sharedArtists = intersect(viewer.recentArtistIds, candidate.recentArtistIds);
  const viewerRecent = viewer.recentTrackIds.length + viewer.recentArtistIds.length;
  const candidateRecent = candidate.recentTrackIds.length + candidate.recentArtistIds.length;
  const denom = Math.min(viewerRecent, candidateRecent);
  const ratio = denom <= 0
    ? 0
    : Math.min(1, (sharedTracks.length + sharedArtists.length) / denom);
  return {ratio, sharedTracks, sharedArtists};
}

function musicBand(score: number): MusicInsightCode {
  if (score >= 85) return "band_very_high";
  if (score >= 70) return "band_high";
  if (score >= 40) return "band_mid";
  return "band_low";
}

export function buildMusicInsights(input: {
  score: number;
  sharedTracks: string[];
  sharedArtists: string[];
  sharedGenres: string[];
  sharedRecentTracks: string[];
  sharedPlaylistTracks: string[];
  /** Display names when resolved; otherwise IDs may appear. */
  trackNames?: string[];
  artistNames?: string[];
}): MusicInsight[] {
  const insights: MusicInsight[] = [{code: musicBand(input.score)}];
  if (input.sharedTracks.length > 0) {
    insights.push({
      code: "shared_tracks",
      params: {count: input.sharedTracks.length},
    });
  }
  if (input.sharedArtists.length > 0) {
    insights.push({
      code: "shared_artists",
      params: {count: input.sharedArtists.length},
    });
  }
  if (input.sharedPlaylistTracks.length > 0) {
    insights.push({
      code: "shared_playlist_tracks",
      params: {count: input.sharedPlaylistTracks.length},
    });
  }
  if (input.sharedRecentTracks.length > 0) {
    insights.push({
      code: "shared_recent_tracks",
      params: {count: input.sharedRecentTracks.length},
    });
  }
  const topArtist = (input.artistNames ?? input.sharedArtists)[0];
  if (topArtist) {
    insights.push({
      code: "top_shared_artist",
      params: {name: topArtist},
    });
  }
  if (input.sharedGenres.length > 0) {
    const genres = input.sharedGenres.slice(0, 2).join(", ");
    insights.push({
      code: "top_shared_genres",
      params: {genres},
    });
  }
  if (
    input.sharedTracks.length === 0 &&
    input.sharedArtists.length === 0 &&
    input.sharedGenres.length === 0 &&
    input.sharedPlaylistTracks.length === 0 &&
    input.sharedRecentTracks.length === 0
  ) {
    insights.push({code: "data_unavailable"});
  }
  return insights;
}

export function scoreMusicCompatibility(
  viewer: MusicTaste,
  candidate: MusicTaste,
): MusicCompatibility {
  if (isTasteEmpty(viewer) || isTasteEmpty(candidate)) {
    return {
      score: 0,
      sharedTracks: [],
      sharedArtists: [],
      sharedGenres: [],
      sharedRecentTracks: [],
      sharedPlaylistTracks: [],
      breakdown: {tracks: 0, artists: 0, genres: 0, recent: 0, playlist: 0},
      insights: [{code: "data_unavailable"}],
    };
  }

  const sharedTracks = intersect(viewer.trackIds, candidate.trackIds);
  const sharedArtists = intersect(viewer.artistIds, candidate.artistIds);
  const sharedGenres = intersect(viewer.genres, candidate.genres);
  const recent = recentOverlap(viewer, candidate);
  const viewerPlaylist = viewer.playlistTrackIds ?? [];
  const candidatePlaylist = candidate.playlistTrackIds ?? [];
  const hasPlaylistSignal = viewerPlaylist.length > 0 && candidatePlaylist.length > 0;
  const sharedPlaylistTracks = hasPlaylistSignal
    ? intersect(viewerPlaylist, candidatePlaylist)
    : [];

  const trackRatio = overlap(viewer.trackIds, candidate.trackIds);
  const artistRatio = overlap(viewer.artistIds, candidate.artistIds);
  const genreRatio = overlap(viewer.genres, candidate.genres);
  const playlistRatio = hasPlaylistSignal
    ? overlap(viewerPlaylist, candidatePlaylist)
    : 0;

  const weights = hasPlaylistSignal ? MUSIC_WEIGHTS_WITH_PLAYLIST : MUSIC_WEIGHTS_BASE;
  const weighted =
    trackRatio * weights.tracks +
    artistRatio * weights.artists +
    genreRatio * weights.genres +
    recent.ratio * weights.recent +
    playlistRatio * weights.playlist;

  const score = Math.max(0, Math.min(100, Math.round(weighted * 100)));
  const breakdown: MusicCompatibilityBreakdown = {
    tracks: Math.round(trackRatio * 100),
    artists: Math.round(artistRatio * 100),
    genres: Math.round(genreRatio * 100),
    recent: Math.round(recent.ratio * 100),
    playlist: Math.round(playlistRatio * 100),
  };

  return {
    score,
    sharedTracks,
    sharedArtists,
    sharedGenres,
    sharedRecentTracks: recent.sharedTracks,
    sharedPlaylistTracks,
    breakdown,
    insights: buildMusicInsights({
      score,
      sharedTracks,
      sharedArtists,
      sharedGenres,
      sharedRecentTracks: recent.sharedTracks,
      sharedPlaylistTracks,
    }),
  };
}

export function musicRankingBonus(score: number): number {
  if (score <= 0) return 0;
  return Math.max(0, Math.min(15, Math.round(score * MUSIC_RANKING_WEIGHT)));
}

export function interestedInAllows(interestedIn: unknown, gender: unknown): boolean {
  const raw = typeof interestedIn === "string" ? interestedIn.trim().toLowerCase() : "";
  if (!raw || raw === "everyone") return true;
  // Settings UI may store singular preferredGender ("man"/"woman").
  const want =
    raw === "man" || raw === "male" || raw === "men"
      ? "men"
      : raw === "woman" || raw === "female" || raw === "women"
        ? "women"
        : raw;
  const g = typeof gender === "string" ? gender.trim().toLowerCase() : "";
  if (!g) return true;
  if (want === "men") return g === "man" || g === "male" || g === "men";
  if (want === "women") return g === "woman" || g === "female" || g === "women";
  return true;
}

export function isoWeekId(date = new Date()): string {
  const utc = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const day = utc.getUTCDay() || 7;
  utc.setUTCDate(utc.getUTCDate() + 4 - day);
  const yearStart = new Date(Date.UTC(utc.getUTCFullYear(), 0, 1));
  const week = Math.ceil(((utc.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
  return `${utc.getUTCFullYear()}-W${String(week).padStart(2, "0")}`;
}

export type NamedMusicItem = {id: string; name: string; artist?: string};

/** Resolve Spotify IDs to display names from both profiles' cached metadata. */
export function resolveMusicNames(
  ids: string[],
  catalogs: NamedMusicItem[],
): string[] {
  const byId = new Map<string, string>();
  for (const item of catalogs) {
    if (!item.id || !item.name) continue;
    byId.set(item.id, item.name);
  }
  const out: string[] = [];
  const seen = new Set<string>();
  for (const id of ids) {
    const name = byId.get(id) ?? "";
    if (!name || seen.has(norm(name))) continue;
    seen.add(norm(name));
    out.push(name);
  }
  return out;
}

export function enrichMusicCompatibility(
  result: MusicCompatibility,
  catalogs: NamedMusicItem[],
): MusicCompatibility & {
  sharedTrackNames: string[];
  sharedArtistNames: string[];
} {
  const sharedTrackNames = resolveMusicNames(result.sharedTracks, catalogs);
  const sharedArtistNames = resolveMusicNames(result.sharedArtists, catalogs);
  return {
    ...result,
    sharedTrackNames,
    sharedArtistNames,
    insights: buildMusicInsights({
      score: result.score,
      sharedTracks: result.sharedTracks,
      sharedArtists: result.sharedArtists,
      sharedGenres: result.sharedGenres,
      sharedRecentTracks: result.sharedRecentTracks,
      sharedPlaylistTracks: result.sharedPlaylistTracks,
      trackNames: sharedTrackNames,
      artistNames: sharedArtistNames,
    }),
  };
}

File: lib/features/music/domain/services/music_compatibility.dart
import 'package:mevora/features/music/domain/entities/music_taste.dart';

enum MusicCompatibilityBand { low, mid, high, veryHigh }

enum MusicInsightCode {
  bandVeryHigh,
  bandHigh,
  bandMid,
  bandLow,
  sharedTracks,
  sharedArtists,
  sharedPlaylistTracks,
  sharedRecentTracks,
  topSharedArtist,
  topSharedGenres,
  dataUnavailable,
  notConnected,
}

class MusicInsight {
  const MusicInsight({
    required this.code,
    this.params = const {},
  });

  final MusicInsightCode code;
  final Map<String, Object> params;
}

class MusicCompatibilityBreakdown {
  const MusicCompatibilityBreakdown({
    this.tracks = 0,
    this.artists = 0,
    this.genres = 0,
    this.recent = 0,
    this.playlist = 0,
  });

  final int tracks;
  final int artists;
  final int genres;
  final int recent;
  final int playlist;
}

class MusicCompatibilityResult {
  const MusicCompatibilityResult({
    required this.score,
    this.sharedTracks = const [],
    this.sharedArtists = const [],
    this.sharedGenres = const [],
    this.sharedRecentTracks = const [],
    this.sharedPlaylistTracks = const [],
    this.sharedTrackNames = const [],
    this.sharedArtistNames = const [],
    this.breakdown = const MusicCompatibilityBreakdown(),
    this.insights = const [],
  });

  /// 0–100 inclusive. Extra discovery signal — never an automatic match.
  final int score;
  final List<String> sharedTracks;
  final List<String> sharedArtists;
  final List<String> sharedGenres;
  final List<String> sharedRecentTracks;
  final List<String> sharedPlaylistTracks;
  final List<String> sharedTrackNames;
  final List<String> sharedArtistNames;
  final MusicCompatibilityBreakdown breakdown;
  final List<MusicInsight> insights;

  MusicCompatibilityBand get band => MusicCompatibilityCalculator.band(score);

  bool get hasSignal => score > 0;

  /// Legacy English reasons kept for older callers; prefer [insights] + l10n.
  List<String> get reasons => insights
      .where((item) => item.code == MusicInsightCode.topSharedArtist)
      .map((item) => 'You both listen to ${item.params['name']}')
      .toList();
}

/// Shared tracks / artists / genres / recent / optional playlist overlap.
/// Mirrors `functions/src/musicCompatibility.ts` for client-side previews.
abstract final class MusicCompatibilityCalculator {
  static const trackWeight = 0.40;
  static const artistWeight = 0.30;
  static const genreWeight = 0.20;
  static const recentWeight = 0.10;

  static const trackWeightWithPlaylist = 0.35;
  static const artistWeightWithPlaylist = 0.25;
  static const genreWeightWithPlaylist = 0.20;
  static const playlistWeight = 0.10;
  static const recentWeightWithPlaylist = 0.10;

  /// Ranking-only bonus so music never replaces dating compatibility.
  static const rankingWeight = 0.15;

  static MusicCompatibilityBand band(int score) {
    if (score >= 85) {
      return MusicCompatibilityBand.veryHigh;
    }
    if (score >= 70) {
      return MusicCompatibilityBand.high;
    }
    if (score >= 40) {
      return MusicCompatibilityBand.mid;
    }
    return MusicCompatibilityBand.low;
  }

  static int rankingBonus(int musicScore) {
    if (musicScore <= 0) {
      return 0;
    }
    return (musicScore * rankingWeight).round().clamp(0, 15);
  }

  static MusicCompatibilityResult score({
    required MusicTasteSnapshot viewer,
    required MusicTasteSnapshot candidate,
    List<String> trackNames = const [],
    List<String> artistNames = const [],
  }) {
    if (viewer.isEmpty || candidate.isEmpty) {
      return const MusicCompatibilityResult(
        score: 0,
        insights: [MusicInsight(code: MusicInsightCode.dataUnavailable)],
      );
    }

    final sharedTracks = _intersect(viewer.trackIds, candidate.trackIds);
    final sharedArtists = _intersect(viewer.artistIds, candidate.artistIds);
    final sharedGenres = _intersect(viewer.genres, candidate.genres);
    final recentTracks = _intersect(
      viewer.recentTrackIds,
      candidate.recentTrackIds,
    );
    final recentArtists = _intersect(
      viewer.recentArtistIds,
      candidate.recentArtistIds,
    );
    final hasPlaylist =
        viewer.playlistTrackIds.isNotEmpty &&
        candidate.playlistTrackIds.isNotEmpty;
    final sharedPlaylist = hasPlaylist
        ? _intersect(viewer.playlistTrackIds, candidate.playlistTrackIds)
        : const <String>[];

    final tracks = _overlap(viewer.trackIds, candidate.trackIds);
    final artists = _overlap(viewer.artistIds, candidate.artistIds);
    final genres = _overlap(viewer.genres, candidate.genres);
    final recent = _combinedRecentOverlap(
      recentTracks: recentTracks.length,
      recentArtists: recentArtists.length,
      viewerRecent:
          viewer.recentTrackIds.length + viewer.recentArtistIds.length,
      candidateRecent:
          candidate.recentTrackIds.length + candidate.recentArtistIds.length,
    );
    final playlist = hasPlaylist
        ? _overlap(viewer.playlistTrackIds, candidate.playlistTrackIds)
        : 0.0;

    final weighted = hasPlaylist
        ? tracks * trackWeightWithPlaylist +
              artists * artistWeightWithPlaylist +
              genres * genreWeightWithPlaylist +
              playlist * playlistWeight +
              recent * recentWeightWithPlaylist
        : tracks * trackWeight +
              artists * artistWeight +
              genres * genreWeight +
              recent * recentWeight;

    final scoreValue = (weighted * 100).round().clamp(0, 100);
    final resolvedTrackNames = trackNames.isNotEmpty
        ? trackNames
        : sharedTracks;
    final resolvedArtistNames = artistNames.isNotEmpty
        ? artistNames
        : sharedArtists;

    return MusicCompatibilityResult(
      score: scoreValue,
      sharedTracks: sharedTracks,
      sharedArtists: sharedArtists,
      sharedGenres: sharedGenres,
      sharedRecentTracks: recentTracks,
      sharedPlaylistTracks: sharedPlaylist,
      sharedTrackNames: resolvedTrackNames,
      sharedArtistNames: resolvedArtistNames,
      breakdown: MusicCompatibilityBreakdown(
        tracks: (tracks * 100).round(),
        artists: (artists * 100).round(),
        genres: (genres * 100).round(),
        recent: (recent * 100).round(),
        playlist: (playlist * 100).round(),
      ),
      insights: buildInsights(
        score: scoreValue,
        sharedTrackCount: sharedTracks.length,
        sharedArtistCount: sharedArtists.length,
        sharedGenreNames: sharedGenres,
        sharedRecentTrackCount: recentTracks.length,
        sharedPlaylistTrackCount: sharedPlaylist.length,
        topArtistName: resolvedArtistNames.isEmpty
            ? null
            : resolvedArtistNames.first,
      ),
    );
  }

  static List<MusicInsight> buildInsights({
    required int score,
    required int sharedTrackCount,
    required int sharedArtistCount,
    required List<String> sharedGenreNames,
    required int sharedRecentTrackCount,
    required int sharedPlaylistTrackCount,
    String? topArtistName,
  }) {
    final insights = <MusicInsight>[
      MusicInsight(code: _bandCode(score)),
    ];
    if (sharedTrackCount > 0) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.sharedTracks,
          params: {'count': sharedTrackCount},
        ),
      );
    }
    if (sharedArtistCount > 0) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.sharedArtists,
          params: {'count': sharedArtistCount},
        ),
      );
    }
    if (sharedPlaylistTrackCount > 0) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.sharedPlaylistTracks,
          params: {'count': sharedPlaylistTrackCount},
        ),
      );
    }
    if (sharedRecentTrackCount > 0) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.sharedRecentTracks,
          params: {'count': sharedRecentTrackCount},
        ),
      );
    }
    if (topArtistName != null && topArtistName.isNotEmpty) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.topSharedArtist,
          params: {'name': topArtistName},
        ),
      );
    }
    if (sharedGenreNames.isNotEmpty) {
      insights.add(
        MusicInsight(
          code: MusicInsightCode.topSharedGenres,
          params: {'genres': sharedGenreNames.take(2).join(', ')},
        ),
      );
    }
    if (sharedTrackCount == 0 &&
        sharedArtistCount == 0 &&
        sharedGenreNames.isEmpty &&
        sharedPlaylistTrackCount == 0 &&
        sharedRecentTrackCount == 0) {
      insights.add(
        const MusicInsight(code: MusicInsightCode.dataUnavailable),
      );
    }
    return insights;
  }

  static MusicInsightCode parseInsightCode(String? raw) {
    return switch (raw) {
      'band_very_high' => MusicInsightCode.bandVeryHigh,
      'band_high' => MusicInsightCode.bandHigh,
      'band_mid' => MusicInsightCode.bandMid,
      'band_low' => MusicInsightCode.bandLow,
      'shared_tracks' => MusicInsightCode.sharedTracks,
      'shared_artists' => MusicInsightCode.sharedArtists,
      'shared_playlist_tracks' => MusicInsightCode.sharedPlaylistTracks,
      'shared_recent_tracks' => MusicInsightCode.sharedRecentTracks,
      'top_shared_artist' => MusicInsightCode.topSharedArtist,
      'top_shared_genres' => MusicInsightCode.topSharedGenres,
      'not_connected' => MusicInsightCode.notConnected,
      _ => MusicInsightCode.dataUnavailable,
    };
  }

  static MusicInsightCode _bandCode(int score) {
    final band = MusicCompatibilityCalculator.band(score);
    return switch (band) {
      MusicCompatibilityBand.veryHigh => MusicInsightCode.bandVeryHigh,
      MusicCompatibilityBand.high => MusicInsightCode.bandHigh,
      MusicCompatibilityBand.mid => MusicInsightCode.bandMid,
      MusicCompatibilityBand.low => MusicInsightCode.bandLow,
    };
  }

  static double _overlap(List<String> a, List<String> b) {
    if (a.isEmpty || b.isEmpty) {
      return 0;
    }
    final shared = _intersect(a, b).length;
    final denom = a.length < b.length ? a.length : b.length;
    if (denom == 0) {
      return 0;
    }
    return (shared / denom).clamp(0, 1);
  }

  static double _combinedRecentOverlap({
    required int recentTracks,
    required int recentArtists,
    required int viewerRecent,
    required int candidateRecent,
  }) {
    final denom = viewerRecent < candidateRecent
        ? viewerRecent
        : candidateRecent;
    if (denom <= 0) {
      return 0;
    }
    return ((recentTracks + recentArtists) / denom).clamp(0, 1);
  }

  static List<String> _intersect(List<String> a, List<String> b) {
    final other = b.map(_norm).where((item) => item.isNotEmpty).toSet();
    final seen = <String>{};
    final out = <String>[];
    for (final item in a) {
      final key = _norm(item);
      if (key.isEmpty || !other.contains(key) || !seen.add(key)) {
        continue;
      }
      out.add(item.trim());
    }
    return out;
  }

  static String _norm(String value) => value.trim().toLowerCase();
}
